docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,323 @@
+    - How much damage is caused by the the version mismatch
+    - The difficulty of debugging a version mismatch
+* Developer experience, it should be easier to write and maintain a new binding generator than it currently is.
+* Releases, cutting releases for changes in one binding generator shouldn't harm another.
+
+## Considered Options
+
+* **[Option 1] Do nothing**
+
+    This means keeping everything as-is, and deciding that all binding generators (at least for now) should live under the `uniffi_bindgen` crate.
+
+* **[Option 2] Publish a separate crate per binding generator and have users use the  `uniffi_bindgen` Command line interface to shell out to the specific binding generators**
+
+    Users would have to install a binary of the crates of the binding language they would like to use **in addition** to installing `uniffi_bindgen`. For example, a user would need to:
+
+    1. run `cargo install uniffi_bindgen_kotlin`.
+    1. run `cargo install uniffi_bindgen`.
+    1. Then a user will finally be able to run `uniffi_bindgen generate -l kotlin <ARGS>`. Users must install `uniffi_bindgen_kotlin` otherwise `uniffi_bindgen` would fail.
+
+    This means that `uniffi_bindgen` would still handle generic tasks related to binding generation. This includes:
+
+    1. Parsing the UDL, and passing the `ComponentInterface` to the specific binding generator.
+    1. Parsing the `uniffi.toml` and passing the language specific configuration down to the binding generator.
+    1. Calling a generic `BindingGenerator::write_bindings` function that is implemented by the specific binding generator.
+    1. `uniffi_bindgen` would also expose its `CodeOracle`, `CodeType` and `CodeDeclaration` types to help developers create a standard way to interact with code generation. It shouldn't however, restrict developers to use those concepts if they don't fit a specific binding.
+    But developers would still need a separate crate that implements the traits `uniffi_bindgen` exposes.
+
+* **[Option 3] Publish a separate crate per binding generator and have them get consumed directly**
+
+    Users would only need to install `uniffi_bindgen_{lang}` and run it like `uniffi_bindgen_kotlin <ARGS>`, and each `uniffi_bindgen_{lang}` could opt to support generating scaffolding as well.
+## Pros and Cons of the Options
+
+### **[Option 1] Do nothing**
+
+* Good, because it makes it hard to accidentally use different versions of `uniffi` for scaffolding and bindings.
+* Good, it makes it easier to make changes to multiple bindings at a time (in the case of a breaking change in `uniffi`, etc).
+* Bad, because maintainability can grow to be difficult - especially if more bindings are added which the core `uniffi` team is not familiar with.
+* Bad, because testability can also grow to be difficult - as more bindings are added (some type of `gecko-js` is expected in the future) the requirement to test all the bindings together in one repository is difficult to maintain.
```

Comment:
Building is hard too because you need the entire build system in place to build anything.